### PR TITLE
Add lr scheduler, weight decay and max_grad_norm

### PR DESCRIPTION
**Pull request by azahed98**
*Have you read the [Contributing Guidelines](https://github.com/togethercomputer/together/blob/main/CONTRIBUTING.md)?*

Issue #

## Describe your changes

*Clearly and concisely describe what's in this pull request. Include screenshots, if necessary.*

### CODE REVIEW

File: src/together/resources/finetune.py [hunk 1]
```diff
@@ -82,6 +87,20 @@ def createFinetuneRequest(
     if warmup_ratio > 1 or warmup_ratio < 0:
         raise ValueError("Warmup ratio should be between 0 and 1")
 
+    if min_lr_ratio is not None and (min_lr_ratio > 1 or min_lr_ratio < 0):
+        raise ValueError("Ending rate should be between 0 and 1")
```
Comment (mryab) [inline src/together/resources/finetune.py:91]:
```suggestion
        raise ValueError("Min learning rate ratio should be between 0 and 1")
```

File: src/together/resources/finetune.py [hunk 2]
```diff
@@ -140,7 +165,11 @@ def create(
             batch_size (int, optional): Batch size for fine-tuning. Defaults to max.
             learning_rate (float, optional): Learning rate multiplier to use for training
                 Defaults to 0.00001.
+            min_lr_ratio (float, optional): Ending learning rate as a percentage of the initial learning rate for
```
Comment (mryab) [inline src/together/resources/finetune.py:168]:
```suggestion
            min_lr_ratio (float, optional): Final learning rate as a percentage of the initial learning rate for
```

File: src/together/resources/finetune.py [hunk 3]
```diff
@@ -140,7 +165,11 @@ def create(
...
+                learning rate scheduler. Defaults to 0.0.
```
Comment (mryab) [inline src/together/resources/finetune.py:169]:
```suggestion
                the learning rate scheduler. Defaults to 0.0.
```

File: src/together/resources/finetune.py [hunk 4]
```diff
@@ -140,7 +165,11 @@ def create(
...
             warmup_ratio (float, optional): Warmup ratio for learning rate scheduler.
+            max_grad_norm (float, optional): Max gradient norm. Defaults to None.
+            weight_decay (float, optional): Weight decay. Defaults to None.
```
Comment (mryab) [inline src/together/resources/finetune.py:172]: Should we allow it to be None? Probably safer to only permit floating point values and use 0 by default

File: src/together/resources/finetune.py [hunk 5]
```diff
@@ -140,7 +165,11 @@ def create(
             batch_size (int, optional): Batch size for fine-tuning. Defaults to max.
             learning_rate (float, optional): Learning rate multiplier to use for training
                 Defaults to 0.00001.
+            min_lr_ratio (float, optional): Ending learning rate as a percentage of the initial learning rate for
+                learning rate scheduler. Defaults to 0.0.
             warmup_ratio (float, optional): Warmup ratio for learning rate scheduler.
+            max_grad_norm (float, optional): Max gradient norm. Defaults to None.
```
Comment (mryab) [inline src/together/resources/finetune.py:171]:
```suggestion
            max_grad_norm (float, optional): Max gradient norm to use for gradient clipping. Defaults to None (no gradient clipping).
```

File: src/together/resources/finetune.py [hunk 6]
```diff
@@ -449,9 +484,13 @@ async def create(
             n_checkpoints (int, optional): Number of checkpoints to save during fine-tuning.
                 Defaults to 1.
             batch_size (int, optional): Batch size for fine-tuning. Defaults to max.
-            learning_rate (float, optional): Learning rate multiplier to use for training
-                Defaults to 0.00001.
+            min_lr_ratio (float, optional): Ending learning rate as a percentage of the initial learning rate for
+                learning rate scheduler. Defaults to 0.0.
+            min_lr_ratio (float, optional): Ending learning rate for learning rate scheduler.
+                Defaults to 0.0.
             warmup_ratio (float, optional): Warmup ratio for learning rate scheduler.
```
Comment (mryab) [inline src/together/resources/finetune.py:491]: I think you accidentally removed the learning rate argument from here. Also, the comments for `FineTuning.create` apply here

File: src/together/resources/finetune.py [hunk 7]
```diff
@@ -91,7 +110,10 @@ def createFinetuneRequest(
         n_checkpoints=n_checkpoints,
         batch_size=batch_size,
         learning_rate=learning_rate,
+        lr_scheduler=lrScheduler,
         warmup_ratio=warmup_ratio,
+        max_grad_norm=max_grad_norm,
+        weight_decay=weight_decay,
```
Comment (mryab) [inline src/together/resources/finetune.py:116]: FinetuneRequest does not seem to have those fields, should we add them?

File: src/together/resources/finetune.py [hunk 8]
```diff
@@ -449,9 +484,13 @@ async def create(
             n_checkpoints (int, optional): Number of checkpoints to save during fine-tuning.
                 Defaults to 1.
             batch_size (int, optional): Batch size for fine-tuning. Defaults to max.
-            learning_rate (float, optional): Learning rate multiplier to use for training
-                Defaults to 0.00001.
+            min_lr_ratio (float, optional): Ending learning rate as a percentage of the initial learning rate for
+                learning rate scheduler. Defaults to 0.0.
+            min_lr_ratio (float, optional): Ending learning rate for learning rate scheduler.
```
Comment (artek0chumak) [inline src/together/resources/finetune.py:489]: You have a repetion here.

File: src/together/cli/api/finetune.py [hunk 9]
```diff
@@ -60,12 +60,30 @@ def fine_tuning(ctx: click.Context) -> None:
 )
 @click.option("--batch-size", type=INT_WITH_MAX, default="max", help="Train batch size")
 @click.option("--learning-rate", type=float, default=1e-5, help="Learning rate")
+@click.option(
+    "--min-lr-ratio",
+    type=float,
+    default=0.0,
+    help="Final learning rate as a percentage of the initial learning rate",
```
Comment (thepowerfuldeez) [inline src/together/cli/api/finetune.py:67]: Is it ratio (0.05 means 5%) or percentage? Or value for min learning rate? Min lr would be more concise in my opinion, but we could specify helpstring somewhat differently I believe

Comment (mryab) [reply src/together/cli/api/finetune.py:67]: I was also confused about exposing the ratio instead of the absolute LR — this seems to be how optimizer schedules are frequently configured (even in Axolotl, there is `cosine_min_lr_ratio` as a config option). However, I agree that "percentage" is misleading, because this ratio goes from 0 to 1

File: src/together/resources/finetune.py (hunk 4 above)
Comment (azahed98) [reply src/together/resources/finetune.py]: Updated to only float

File: src/together/resources/finetune.py (hunk 6 above)
Comment (azahed98) [reply src/together/resources/finetune.py]: Sorry, added back and updated.

File: src/together/resources/finetune.py (hunk 7 above)
Comment (azahed98) [reply src/together/resources/finetune.py:116]: Ty, updated

File: src/together/cli/api/finetune.py (hunk 9 above)
Comment (azahed98) [reply src/together/cli/api/finetune.py]: Good point, removed "percentage" here and will update in the together-finetune PR as well

File: src/together/cli/api/finetune.py [hunk 10]
```diff
@@ -65,12 +65,30 @@ def fine_tuning(ctx: click.Context) -> None:
 )
 @click.option("--batch-size", type=INT_WITH_MAX, default="max", help="Train batch size")
 @click.option("--learning-rate", type=float, default=1e-5, help="Learning rate")
+@click.option(
+    "--min-lr-ratio",
+    type=float,
+    default=0.0,
+    help="Final learning rate ratio of the initial learning rate",
```
Comment (mryab) [inline src/together/cli/api/finetune.py:72]:
```suggestion
    help="The ratio of the final learning rate to the peak learning rate",
```

File: src/together/cli/api/finetune.py [hunk 11]
```diff
@@ -65,12 +65,30 @@ def fine_tuning(ctx: click.Context) -> None:
...
     "--warmup-ratio",
     type=float,
     default=0.0,
     help="Warmup ratio for learning rate scheduler.",
 )
+@click.option(
+    "--max-grad-norm",
+    type=float,
+    default=1.0,
+    help="Max gradient norm. Set to 0 to disable.",
```
Comment (mryab) [inline src/together/cli/api/finetune.py:84]:
```suggestion
    help="Max gradient norm to be used for gradient clipping. Set to 0 to disable.",
```

File: src/together/resources/finetune.py [hunk 12]
```diff
@@ -35,7 +37,10 @@ def createFinetuneRequest(
     n_checkpoints: int | None = 1,
     batch_size: int | Literal["max"] = "max",
     learning_rate: float | None = 0.00001,
+    min_lr_ratio: float | None = 0.0,
```
Comment (mryab) [inline src/together/resources/finetune.py:40]: Can we define it simply as float, since this should be always defined by that point?

File: src/together/resources/finetune.py [hunk 13]
```diff
@@ -143,7 +168,11 @@ def create(
             batch_size (int or "max"): Batch size for fine-tuning. Defaults to max.
             learning_rate (float, optional): Learning rate multiplier to use for training
                 Defaults to 0.00001.
+            min_lr_ratio (float, optional): Min learning rate ratio of the initial learning rate for
+                the learning rate scheduler. Defaults to 0.0.
             warmup_ratio (float, optional): Warmup ratio for learning rate scheduler.
+            max_grad_norm (float, optional): Max gradient norm. Defaults to 1.0. Set to 0 to disable.
```
Comment (mryab) [inline src/together/resources/finetune.py:174]:
```suggestion
            max_grad_norm (float, optional): Max gradient norm. Defaults to 1.0, set to 0 to disable gradient clipping.
```

File: src/together/resources/finetune.py [hunk 14]
```diff
@@ -462,7 +497,11 @@ async def create(
             batch_size (int, optional): Batch size for fine-tuning. Defaults to max.
             learning_rate (float, optional): Learning rate multiplier to use for training
                 Defaults to 0.00001.
+            min_lr_ratio (float, optional): Min learning rate ratio of the initial learning rate for
+                the learning rate scheduler. Defaults to 0.0.
             warmup_ratio (float, optional): Warmup ratio for learning rate scheduler.
+            max_grad_norm (float, optional): Max gradient norm. Defaults to 1.0. Set to 0 to disable.
```
Comment (mryab) [inline src/together/resources/finetune.py:503]:
```suggestion
            max_grad_norm (float, optional): Max gradient norm. Defaults to 1.0, set to 0 to disable gradient clipping.
```

File: src/together/resources/finetune.py [hunk 15]
```diff
@@ -436,7 +468,10 @@ async def create(
         n_checkpoints: int | None = 1,
         batch_size: int | Literal["max"] = "max",
         learning_rate: float | None = 0.00001,
+        min_lr_ratio: float | None = 0.0,
```
Comment (mryab) [inline src/together/resources/finetune.py:471]: Same here

File: src/together/resources/finetune.py [hunk 16]
```diff
@@ -117,7 +139,10 @@ def create(
...
```
Comment (mryab) [inline src/together/resources/finetune.py:142]: Is this also always a float now?

File: src/together/types/finetune.py [hunk 17]
```diff
@@ -150,8 +150,14 @@ class FinetuneRequest(BaseModel):
...
     # training learning rate
     learning_rate: float
+    # learning rate scheduler type and args
+    lr_scheduler: FinetuneLRScheduler | None = None
     # learning rate warmup ratio
     warmup_ratio: float
+    # max gradient norm
+    max_grad_norm: float | None = None
+    # weight decay
+    weight_decay: float | None = None
```
Comment (mryab) [inline src/together/types/finetune.py:160]: I understand why it can be None for responses (for example, older jobs without the argument), but can it be just `float` in case of requests?

File: src/together/resources/finetune.py (hunk 12 above)
Comment (azahed98) [reply src/together/resources/finetune.py]: Done. Also changed for max_grad_norm and weight_decay

File: src/together/types/finetune.py (hunk 17 above)
Comment (azahed98) [reply src/together/types/finetune.py]: Good point, changed!

**Review by mryab — APPROVED** [review]
